# Film studio dataset

By this point I have already gathered some tweets related to names of certain film studios. Now I have to explore what I actually have, and do some sort of filtration of it. For the last part I wanted to try and label the data.

In [1]:
#collapse
import os
import numpy as np
import pandas as pd
import glob

In [2]:
#collapse
paths=[
    ("Dreamworks", r"..\datasets\films\DreamWorks_5_2000.csv"),
    ("Disney", r"..\datasets\films\disney_5_2000.csv"),
    ("Marvel", r"..\datasets\films\marvel_5_2000.csv"),
    ("Studio Ghibli", r"..\datasets\films\studioGhibli_5_2000.csv"),
    ("Sony Pictures", r"..\datasets\films\SonyPictures_5_2000.csv"),
    ("Warner Bros", r"..\datasets\films\WarnerBros._5_2000.csv"),
    ("DC", r"..\datasets\films\dc_5_2000.csv"),
    ("Pixar", r"..\datasets\films\pixar_5_2000.csv"),
    ("Lionsgate", r"..\datasets\films\Lionsgate_5_2000.csv")
]

li = []

for studio, path in paths:
    df = pd.read_csv(path, index_col=0, header=0)
    df["studio"] = studio
    li.append(df)

df_studios = pd.concat(li, axis=0, ignore_index=True)

In [3]:
df_studios.sample(5)

username              name  \
2282   DisneyHotelBot     ディズニーホテル空室状況！   
6135  themanoftomorr7  themanoftomorrow   
6248        DCtamilFC       𝐃𝐂 𝐓𝐚𝐦𝐢𝐥 𝐅𝐂   
2271        ma_san100      まぁさん「100の冒険」   
4106      sherwin_lui         Blastpoke   

                                profile_picture  replies  retweets  likes  \
2282   https://twitter.com/DisneyHotelBot/photo        0         0      0   
6135  https://twitter.com/themanoftomorr7/photo        0         1      0   
6248        https://twitter.com/DCtamilFC/photo        2        21     51   
2271        https://twitter.com/ma_san100/photo        0         0      0   
4106      https://twitter.com/sherwin_lui/photo        0         0      0   

      is_retweet                posted_time  \
2282       False  2022-04-24T23:28:07+00:00   
6135       False  2022-04-24T15:11:31+00:00   
6248       False  2022-04-24T12:26:02+00:00   
2271       False  2022-04-24T23:28:16+00:00   
4106       False  2022-04-24T22:50:54+00:00   

                                                content  \
2282  東京ディズニーシー・ホテルミラコスタの\n以下日程が空いています！part2\n\n2022...   
6135  Dylan O'Brien is nightwing!! #Nightwing \n@DCC...   
6248   Warner Bros. Pictures' presentation is on 4pm...   
2271  ウエストポーチ、一目惚れ\n\n#ひとりディズニー のお供にしたい\n\n#天国に持って行く...   
4106  Any char walking across the main room& opp is ...   

                                               hashtags  \
2282                        ['チケット', 'Disney', 'ディズニー']   
6135                                      ['Nightwing']   
6248                          ['TheFlash', 'BlackAdam']   
2271  ['ひとりディズニー', '天国に持って行く思い出', 'ショップディズニー', 'ディズニ...   
4106                                                 []   

                                               mentions  \
2282                                                 []   
6135  ['DCComics', 'warnerbros', 'wbpictures', 'wbd'...   
6248                                      ['CinemaCon']   
2271                                                 []   
4106                                                 []   

                                                 images  \
2282                                                 []   
6135  ['https://pbs.twimg.com/media/FRHk2KdXsAAo2Iy?...   
6248  ['https://pbs.twimg.com/media/FRG_CzxaIAAAdDY?...   
2271                                                 []   
4106                                                 []   

                                                 videos  \
2282                                                 []   
6135                                                 []   
6248                                                 []   
2271                                                 []   
4106  ['https://video.twimg.com/tweet_video/FRI-21gX...   

                                              tweet_url  \
2282  https://twitter.com/DisneyHotelBot/status/1518...   
6135  https://twitter.com/themanoftomorr7/status/151...   
6248  https://twitter.com/DCtamilFC/status/151820461...   
2271  https://twitter.com/ma_san100/status/151837127...   
4106  https://twitter.com/sherwin_lui/status/1518361...   

                         link       studio  
2282  https://t.co/HtmjkaoHYp       Disney  
6135                      NaN  Warner Bros  
6248                      NaN  Warner Bros  
2271  https://t.co/3I7umAwjnv       Disney  
4106                      NaN       Marvel

In [4]:
df_studios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8410 entries, 0 to 8409
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   username         8410 non-null   object
 1   name             8336 non-null   object
 2   profile_picture  8410 non-null   object
 3   replies          8410 non-null   int64 
 4   retweets         8410 non-null   int64 
 5   likes            8410 non-null   int64 
 6   is_retweet       8410 non-null   bool  
 7   posted_time      8410 non-null   object
 8   content          8216 non-null   object
 9   hashtags         8410 non-null   object
 10  mentions         8410 non-null   object
 11  images           8410 non-null   object
 12  videos           8410 non-null   object
 13  tweet_url        8410 non-null   object
 14  link             749 non-null    object
 15  studio           8410 non-null   object
dtypes: bool(1), int64(3), object(12)
memory usage: 993.9+ KB


Tried to download 2000 tweets related to the keywords Dreamworks, Disney, Marvel, Studio Ghibli, Sony Pictures, Warner Bros, DC, Pixar and Lionsgate, which are all related to some film studio. As seen above the dataset consist of only 8410 data points.

The value counts in the cell below show how the data is distributed across the studios. Which shows that some keywords worked well and others did not really work at all.  

In [5]:
df["studio"].value_counts()

Lionsgate    169
Name: studio, dtype: int64

I chose to reduce the dataset to only keep the `content` and `studio` columns.

In [6]:
#collapse
df = df_studios[["content", "studio"]].copy()

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8410 entries, 0 to 8409
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   content  8216 non-null   object
 1   studio   8410 non-null   object
dtypes: object(2)
memory usage: 131.5+ KB


Dropping possible Nan and duplicates.

In [8]:
df.drop_duplicates(inplace=True)
df.apply(lambda x: pd.Series(x.dropna().values))
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8083 entries, 0 to 8409
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   content  8074 non-null   object
 1   studio   8083 non-null   object
dtypes: object(2)
memory usage: 189.4+ KB


In [9]:
df.sample(10)

content         studio
6817  Turning Red - Estranhamente Vermelho no TV Tim...          Pixar
2189                   Disneyland but not Disney World.         Disney
5681  @SonyPictures\n @TomHolland1996\n @Zendaya\n @...  Sony Pictures
7781                                             #pixar          Pixar
3847  Eu gosto de alguns filmes da Marvel, mas acho ...         Marvel
1086  dps de heartstopper eu to esperando ansiosamen...         Disney
6237  The people: Warner Bros, Bring Johnny Depp bac...    Warner Bros
4080  acabei de ver um tweet dc vs marvel e lembrei ...         Marvel
5538  i wanna go to the studio ghibli exhibit so bad...  Studio Ghibli
5536  my beloved friends as iconic studio ghibli sce...  Studio Ghibli

# Detect language

As I was rather awkwardly reminded when working on the "France" dataset, Twitter is a multilingual platform. So for the next part, I will work on detecting which language a tweet is written in. While creating a dataset with multiple languages could be interesting, it is likely to be rather skewed distributed. 

At first, I tried to use TextBlob to detect the language, but it seemed the library had some issues. So I tried to use the langdetect library which said it supports 55 languages. However, trying to detect the language of the tweets directly led to some errors. Mainly an Exception saying "No features in text", which was not something I had stumbled upon in the "France" dataset. 

I tried to do some preprocessing of the text to make the language detection work. First I removed tweets with less than 5 whitespaces. Then I realized that trying to detect the language of anything which does not contain any alphabet did not work, so I tried to remove those.

In [10]:
df = df[df["content"].str.split().str.len().ge(5)]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7315 entries, 0 to 8408
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   content  7315 non-null   object
 1   studio   7315 non-null   object
dtypes: object(2)
memory usage: 171.4+ KB


In [11]:
#collapse
import re

In [12]:
def contains_alphabets(s):
    return bool(re.match("^(?=.*[a-zA-Z])", s))

df["contains_alphabets"] = df["content"].apply(lambda x: contains_alphabets(x))
df["contains_alphabets"].value_counts()

True     7231
False      84
Name: contains_alphabets, dtype: int64

In [13]:
#collapse
df = df[df.contains_alphabets]

However, I still got the same exception. So I tried some other possible solutions: 
- remove punctuation
- remove emoji
- remove html
- remove urls

In [14]:
#collapse
import string

In [16]:
#collapse
def remove_punctuation(text):
    text = text.strip()
    words = text.split()
    table = str.maketrans("", "", string.punctuation)
    words = [word.translate(table) for word in words] # removes Punctuation
    text = " ".join(words)
    return text

def remove_html(text):
    html=re.compile(r"<.*?>")
    return html.sub(r"",text)

def remove_url(text):
    url = re.compile(r"https?://\S+|www\.\S+")
    return url.sub(r"",text)

def remove_emoji(string):
    emoji_pattern = re.compile("["
                               u"\U0001F600-\U0001F64F"  # emoticons
                               u"\U0001F300-\U0001F5FF"  # symbols & pictographs
                               u"\U0001F680-\U0001F6FF"  # transport & map symbols
                               u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                               u"\U00002500-\U00002BEF"  # chinese char
                               u"\U00002702-\U000027B0"
                               u"\U00002702-\U000027B0"
                               u"\U000024C2-\U0001F251"
                               u"\U0001f926-\U0001f937"
                               u"\U00010000-\U0010ffff"
                               u"\u2640-\u2642"
                               u"\u2600-\u2B55"
                               u"\u200d"
                               u"\u23cf"
                               u"\u23e9"
                               u"\u231a"
                               u"\ufe0f"  # dingbats
                               u"\u3030"
                               "]+", flags=re.UNICODE)
    return emoji_pattern.sub(r"", string)

In [17]:
df["pre_text"] = df["content"].apply(lambda x: remove_emoji(x))
df["pre_text"] = df["pre_text"].apply(lambda x: remove_url(x))
df["pre_text"] = df["pre_text"].apply(lambda x: remove_html(x))
df["pre_text"] = df["pre_text"].apply(lambda x: remove_punctuation(x))

But even after having tried those methods, the exception prevailed. So I decided to use the unidecode library to force it to be ASCII. While not being ideal, made the exception go away. Which finally gave me some overview over the detected languages. 

In [18]:
#collapse
import unidecode

In [19]:
#collapse
def make_unidecode(raw_text):
    processed = unidecode.unidecode(raw_text)
    return processed

In [20]:
df["pre_text"] = df["pre_text"].apply(lambda x: make_unidecode(x))

In [21]:
df.sample(10)

content         studio  \
2850  Lost cat Marvel in Rockford, MI US (49341) #lo...         Marvel   
1494  can I say yes?  I have high enough confidence ...         Disney   
6920  Check out this listing I just added to my #Pos...          Pixar   
6575  Not just you, either.\nAnd the Warner Bros. ca...    Warner Bros   
1128  @Disney\n You forgot Disney supports forced in...         Disney   
1975  ...regs, etc. Disney losing it's privs has vir...         Disney   
7809            Os incríveis eh o melhor filme da pixar          Pixar   
7805  PIXAR NO TE VOY A PERDONAR CÓMO TRAUMES A MÍ H...          Pixar   
5321  oh to be one of them studio ghibli girlies who...  Studio Ghibli   
1268  Ron DeSantis’s attack on Disney is obviously u...         Disney   

      contains_alphabets                                           pre_text  
2850                True    Lost cat Marvel in Rockford MI US 49341 lostcat  
1494                True  can I say yes I have high enough confidence to...  
6920                True  Check out this listing I just added to my Posh...  
6575                True  Not just you either And the Warner Bros cartoo...  
1128                True  Disney You forgot Disney supports forced indoc...  
1975                True  regs etc Disney losing its privs has virtually...  
7809                True            Os incriveis eh o melhor filme da pixar  
7805                True  PIXAR NO TE VOY A PERDONAR COMO TRAUMES A MI H...  
5321                True  oh to be one of them studio ghibli girlies who...  
1268                True  Ron DeSantis's attack on Disney is obviously u...

In [22]:
#collapse
from langdetect import detect

In [23]:
df["language"] = df["pre_text"].apply(detect)

In [24]:
df["language"].value_counts()

en    5358
pt     670
es     571
fr     120
it     105
ca      83
id      69
de      57
af      24
et      23
so      17
nl      17
tr      16
da      14
tl      12
no      10
cy       9
ro       8
sl       8
pl       8
sv       5
hr       5
sw       5
hu       4
lt       3
cs       3
fi       3
sk       2
lv       1
vi       1
Name: language, dtype: int64

Since I was not too happy with that solution I tried to use the googletrans library to detect a language. Which was considerably slower, using 20 minutes on something langdetect did in half a minute. Though, it worked on the unaltered text. As the value count shows, goolgetrans detects more languages. Their estimation of how many tweets were English was quite similar. I decided to only go for the rows where they agreed. 

In [25]:
#collapse
from googletrans import Translator

In [26]:
def detected_language(s):
    detector = Translator()
    return detector.detect(s).lang
df["language_content"] = df["content"].apply(lambda x: detected_language(x))

In [27]:
df["language_content"].value_counts()

en             5376
pt              777
es              614
fr              125
id               61
ja               59
it               40
de               28
ca               25
th               19
tr               17
tl               12
pl               10
ru                7
ar                7
nl                4
ko                4
el                4
da                3
su                3
hi                3
bn                2
fi                2
ms                2
[es, pt]          2
no                1
sk                1
si                1
[ms, id]          1
hu                1
[id, jw]          1
[es, ro]          1
[mr, hi]          1
ro                1
ig                1
fa                1
[ja, zh-CN]       1
bg                1
[pt, en]          1
[en, el]          1
af                1
gu                1
[hi, mr]          1
[en, fr]          1
[en, da]          1
cs                1
rw                1
sv                1
te                1
kn                1


In [46]:
#collapse
df_en = df.loc[(df["language"].str.startswith("en")) & (df["language_content"].str.startswith("en"))]

In [48]:
#collapse
df_en = df_en[["content"]].copy()

In [49]:
df_en.sample(10)

content
3099                                                                                                                                                                      Watching No Way Home for the first time since seeing it in theaters. Still the best Marvel film to date
8159                                                                                                                                                                                                             2 days until I can watch Turning Red! #Disney #Pixar #TurningRed
6163                                         Gonna get a platinum trophy as part of my mid year bucket list\n\n#legostarwars #starwars #theforceawakens #howiplaystation #playstation #playstationau #ps4 #bucketlist #starwarstheforceawakens #warnerbros #ttgames #playstation4
3268                                                                                                      I like marvel movies too, just not as much as others. I think true cinema is something that stays with you, not a black and white Albanian film from the pov of a worm.
1497                                                                                                                                    23,95€   Disney Classic Games Collection: The Jungle Book, Aladdin, and The Lion King - NSW https://amzn.to/3EIb11M a través de \n@amazon
6686                                                                                                                                                              @alioverse\n #muchmeow 3D animated NFTs from Artists with previous works done for Disney, Pixar and Dreamworks.
6459                                                                                                                                                                                                             @warnerbros\n @hbomax\n please reconsider the Boondocks revival.
7605         But you can’t make a good argument that superman’s race is integral to the character. Hence we’re seeing diversity in these traditional roles, ie The Zods are black in the Krypton show and Young Justice, and they’re tossing around the idea of a black Superman.
2947  JRJR is still one of the VERY best Spidey artists out there (especially since he seems to be over his tiny ankles phase). Bold cinematic compositions with a dash of JRSR and Kirby still peeping out here and there. JR will always be a bright star in the comics cosmos.
1494                                                                                                   can I say yes?  I have high enough confidence to say that this is a disney princess! you’re not wrong, oppa! and how about you? are you some kind of prince or something?

In [50]:
df_en.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5245 entries, 0 to 8408
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   content  5245 non-null   object
dtypes: object(1)
memory usage: 82.0+ KB


In [51]:
df_en.describe()

content
count                                                                                                                                                                                                                                                                                     5245
unique                                                                                                                                                                                                                                                                                    5170
top     no words. Make it happen. #start #change #now #justiceforjohnnydepp #johnnydepp #istandwithjohnnydepp #justicereform \n@disney\n @warnerbrosentertainment \n@sonypictures\n @paramountpics\n @universalpictures @columbiapictures \n@lionsgate\n @dreamworks\n @theacademy\n @sagaftra
freq                                                                                                                                                                                                                                                                                         3

# Label data

Since the gathered data was unlabeled I wanted to try and explore possible ways of labeling it. Most of my time was used exploring the Snorkel library. Their idea is to create several labeling functions where domain information is injected, and then automatically estimate each function's accuracy and correlation. Then their estimate is combined to determine output labels. However, this relies on having some domain knowledge and an idea of what might indicate each label. Which I do not have, nor wanted to use the time to determine. What I found was that most of their examples with labeling polarity and sentiment relied on a library named TextBlob (the same I tried to use to detect languages). So I decided to explore using that library as a possibility.

> The sentiment property returns a namedtuple of the form Sentiment(polarity, subjectivity). The polarity score is a float within the range [-1.0, 1.0]. The subjectivity is a float within the range [0.0, 1.0] where 0.0 is very objective and 1.0 is very subjective.

So after having extracted the sentiment from the tweet, thresholds are needed to sort them accordingly. 

In [52]:
#collapse
from textblob import TextBlob
from textblob.sentiments import NaiveBayesAnalyzer

In [53]:
#collapse
df_en["score"] = .0 

In [54]:
# Iter throuth the content column and updates the score based on what sentiment score TextBlob gives 
for i, x in df_en.content.iteritems():
    score = TextBlob(x)
    df_en.at[i, "score"] = score.sentiment.polarity

In [55]:
df_en.head()

content  \
0  Oh man, look at Dreamworks, releasing movies with really cool animation styles! This looks amazing! And Trolls 2 had great animation before it too!   
3             Don’t go past this 1/1 \n@alioverse\n #muchmeow 3D animated NFTs from Artists with previous works done for Disney, Pixar and Dreamworks.   
5                                                                                                    Why does he do the dreamworks face in every photo   
6                                                                               i can't believe jerma is the purple man from hit dreamworks movie home   
7                                      @alioverse\n #muchmeow 3D animated NFTs from Artists with previous works done for Disney, Pixar and Dreamworks.   

      score  
0  0.729167  
3 -0.208333  
5  0.000000  
6  0.000000  
7 -0.166667

In [56]:
df_en.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5245 entries, 0 to 8408
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   content  5245 non-null   object 
 1   score    5245 non-null   float64
dtypes: float64(1), object(1)
memory usage: 252.0+ KB


In [57]:
def polarity_to_label(x):
    if(x >= -1 and x < 0):
        return "negative"
    if(x == 0): 
        return "neutral"
    if(x > 0 and x <= 1):
        return "positive"

df_en["label"] = df_en["score"].apply(lambda x: polarity_to_label(x))

In [58]:
df_en["label"].value_counts()

positive    2580
neutral     1699
negative     966
Name: label, dtype: int64

In [59]:
df_en = df_en[["content", "label"]].copy()

In [60]:
#collapse
pd.set_option("display.max_colwidth", None) 

In [61]:
df_en.sample(10)

content  \
2402                                                                                              Disney employs 80,000 Floridians, they most certainly can make a stand to protect their civil rights\n\nAlso makes good business sense - they benefit from a happy pool of potential employees   
6741                                             ok I caught up on another Pixar and yeah that's quite a concept for the mogul's house the family use in The Incredibles 2. not clear what really holds anything up though. also looks like a nightmare for heating in winter, and A/C in summer   
4675                                                                                                                                                                                                                      Maybe not in the cinemas but MARVEL HEROES are definitely in the house   
4808    I took a walk in Koganei, where Studio Ghibli is located. It is an area with many attractions such as a nursery school for the staff's children and Hayao Miyazaki's atelier.\n#studioghibli #hayaomiyazaki #totoro\n\nhttps://mtlnk.net/j_s%253A%252F%252Fwww.bura-bura-tv.com%252F139…   
6055                                                                                                                                                                                                                  And Warner bros large mint projects. They bring new customers to the marky   
8011                                                                                                                                                                                                                                         Also Pixar back in 1988 wasn’t as big as it is now.   
7149                                                                                                                                                                                How sick would it be if Disney/Pixar did a 4*Town virtual concert residency at Walt Disney World in Florida?   
7584                                                                                                                                                                                        by your logic we are canon to the mcu, and so is dc , and pixar, and dreamworks, and illumination...   
3711  Lmao. Making exactly what everyone is asking for is all the time is lame af. That's why the entertainment industry utterly lacks creativity right now. To many focus groups, to much pandering to the market.\n\nMarvel movie after marvel movie, marvel games gonna end up much the same.   
7097                                                                                                                          To infinity and beyond! \nBlast off with our newest legos for “Lightyear”!\n\n\n\n#buzzlightyear #lightyear #disney #pixar #toystory #space #barnesandnoble #142bn   

         label  
2402  positive  
6741  positive  
4675   neutral  
4808  positive  
6055  positive  
8011   neutral  
7149  negative  
7584   neutral  
3711  positive  
7097   neutral

In [62]:
#collapse
df_en.to_csv(r"../datasets/studio_data.csv")